### PR TITLE
Changed Checks from TriviallyCopyable to TriviallyCopyConstructible

### LINKED ISSUE #47355: range-loop-analysis checks for trivial copyability, rather than trivial copy-constructibility
|  |  |
| --- | --- |
| Bugzilla Link | [48011](https://llvm.org/bz48011) |
| Version | 11.0 |
| OS | Linux |
| CC | @dwblaikie,@devincoughlin,@randomnetcat,@pogo59,@zygoloid,@Cry-Thomas |

## Extended Description 
On the Cpplang slack, user Wolle posted a curious warning that came from range-loop-analysis.

Here's a minimal example: https://godbolt.org/z/1cxs43. Note that without the copy assignment operator overload, the warning is not present.

The range-based for loop performs a copy construction, so that is what should be checked for.

At https://clang.llvm.org/doxygen/SemaStmt_8cpp_source.html#l02824, the check is on whether the type is trivially copyable, not whether it is trivially copy constructible, which is a weaker requirement (but the relevant one in this case). I am not intimately familiar with clang sourcecode, but I believe this to be the cause of the warning.

### PR DESCRIPTION
**Pull request by 11happy**
**Overview:**
This pull request fixes #47355 where in the Clang compiler's range-loop-analysis incorrectly checks for trivial copyability instead of trivial copy constructibility, leading to erroneous warnings.

**Changes Made:**
- The changes made in this commit address the issue by introducing a new member function `isTriviallyCopyConstructible` in the `CXXRecordDecl` class and implementing it in the associated files `(clang/include/clang/AST/DeclCXX.h, clang/lib/AST/DeclCXX.cpp)`. Additionally, modifications are made in `clang/include/clang/AST/Type.h` and `clang/lib/AST/Type.cpp` to support the new function. The implementation checks for trivial copy constructibility, distinguishing it from the existing `isTriviallyCopyable` check. The issue in `clang/lib/Sema/SemaStmt.cpp` is also addressed by updating the conditional check in the `DiagnoseForRangeConstVariableCopies` function to use the new `isTriviallyCopyConstructibleType` function. Overall, these changes aim to correct the range-loop-analysis by ensuring it checks for trivial copy constructibility rather than trivial copyability.

**Testing:**
- Tested the updated code.
- Verified that other functionalities remain unaffected.
Screenshot from 2024-01-01 19-50-20

**Dependencies:**
- No dependencies on other pull requests.

**References:**
- https://cplusplus.com/reference/type_traits/is_trivially_copy_constructible/
- https://en.cppreference.com/w/cpp/named_req/CopyConstructible
- https://cplusplus.com/reference/type_traits/is_trivially_copyable/

**CC:**
- @Endilll , @r4nt , @AaronBallman

### CODE REVIEW

**Comment by Endilll** [comment]
Can you add some tests?

**Comment by 11happy** [comment]
Sure I can , I have also tested the code that was giving erroneous warnings, so should I add that example as well??

**Comment by Endilll** [comment]
> Sure I can , I have also tested the code that was giving erroneous warnings, so should I add that example as well??

Yes. Clang has a special mode which check which diagnostics are emitted: https://clang.llvm.org/docs/InternalsManual.html#verifying-diagnostics

**Comment by 11happy** [comment]
I have added the tests as well.

**Comment by 11happy** [comment]
Hello @Endilll Can I work on some other issues also till this PR get reviewed?
Thank you

File: clang/lib/AST/Type.cpp [hunk 1]
```diff
@@ -2644,6 +2644,49 @@ bool QualType::isTriviallyCopyableType(const ASTContext &Context) const {
   return false;
 }
 
+bool QualType::isTriviallyCopyConstructibleType(
```
Comment (cor3ntin) [inline clang/lib/AST/Type.cpp:2647]: Given that this function is exactly the same as the previous one, maybe we could
introduce a (non member static) `isTriviallyCopyableTypeImpl` function that checks either isTriviallyCopyableType or isTriviallyCopyConstructibleType , by either using a boolean or a callable, which isTriviallyCopyConstructibleType and isTriviallyCopyableType would dispatch to

**Comment by 11happy** [comment]
Hello @cor3ntin I have made changes as you suggested.
Thank you

File: clang/lib/AST/Type.cpp [hunk 2]
```diff
@@ -2604,19 +2604,22 @@ bool QualType::isTrivialType(const ASTContext &Context) const {
   return false;
 }
 
-bool QualType::isTriviallyCopyableType(const ASTContext &Context) const {
-  if ((*this)->isArrayType())
-    return Context.getBaseElementType(*this).isTriviallyCopyableType(Context);
+static bool isTriviallyCopyableTypeImpl(const QualType &type,
+                                        const ASTContext &Context,
+                                        bool copy_constructible) {
```
Comment (cor3ntin) [inline clang/lib/AST/Type.cpp:2609]:
```suggestion
                                        bool IsCopyConstructible) {
```

**Review by cor3ntin — COMMENTED** [review]
Looking good beside a small nitpick

**Comment by cor3ntin** [comment]
Thanks.
Can you also add a release note mentioning the issue is fixed? Thanks (clang/docs/ReleaseNotes.rst)

**Comment by 11happy** [comment]
The release notes being quite extensive where should I add it, In bug fixes? or in improvements?

**Comment by cor3ntin** [comment]
In bug fixes is fine, thanks!

**Review by cor3ntin — APPROVED** [review]
LGTM
Make sure to rebase to fix the conflict. 

Thanks for working on this

**Comment by 11happy** [comment]
Thank you @cor3ntin  for your guidance being a new contributor to LLVM your guidance was very helpful and as my First PR gets approved this makes me more excited to work and contribute more to the org.

**Pull request closed without merging**